src/main/kotlin/org/rust/ide/utils/import/importUtils.kt

```diff
@@ -133,6 +127,27 @@ private fun RsUseItem.tryGroupWith(
     return true
 }
 
+private fun tryInsertUseItemAtCorrectLocation(mod: RsItemsOwner, useItem: RsUseItem): Boolean {
+    val newline = RsPsiFactory(mod.project).createNewline()
+    val useWrapper = UseItemWrapper(useItem)
+    val usesAll = mod.childrenOfType<RsUseItem>().map(::UseItemWrapper)
+    if (usesAll.isEmpty()) {
+        val anchor = mod.childrenOfType<RsExternCrateItem>().lastOrNull() ?: return false
+        mod.addBefore(newline, mod.addAfter(useItem, anchor))
+        return true
+    }
+
+    val usesSameLevel = usesAll.filter { it.packageGroupLevel == useWrapper.packageGroupLevel }
+    val uses = usesSameLevel.takeIf { it.isNotEmpty() } ?: usesAll
+    val anchor = uses.takeWhile { it < useWrapper }.lastOrNull()
```

Comment:
What if original imports are not sorted? Does it seem we will find **some** element but not the closest in term of sorting?
How do other languages behave in similar situations?
Maybe just better tocall `val (less, greater) = usesAll.partition { it < useWrapper }` and use `less.lastOrNull()` or `greater.firstOrNull()` as anchor?

Reply:
Anyway, it would be great to have test(s) for such case

Reply:
If original imports are not sorted, then I think it is not that important where we will insert new import (in any way imports remain unsorted). 

E.g. in JavaScript import is inserted before first greater import:
```js
import {func0} from "./foo0";
import {func3} from "./foo3";
import {func1} from "./foo1";
```
After adding import for `func2`:

```js
import {func0} from "./foo0";
import {func2} from "./foo2";
import {func3} from "./foo3";
import {func1} from "./foo1";
```

But I agree that your approach with partition seems to be better, I changed to use it